README
Lingua::Alphabet::Phonetic
==========================

Converts characters or strings to a phonetic alphabet.  
The most common phonetic alphabet is the U.S. Military / NATO standard
where "ABC...YZ" is pronounced "Alpha Bravo Charlie ... Yankee Zulu".
At present, that is the only alphabet available for conversion.


INSTALLATION

To install this module run the following:

   perl Makefile.PL
   make install


COPYRIGHT AND LICENSE

Copyright (C) 2004-2007 Martin Thurn

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

For more information about my Perl Modules, visit http://tinyurl.com/nn67z



Changes
2008-03-05    <Daddy@C-3PO>

	* Kwalitee upgrades

2007-04-21    <Daddy@C-3PO>

	* lib/Lingua/Alphabet/Phonetic.pm: add pod for new()

2006-10-07    <Daddy@C-3PO>

	* lib/Lingua/Alphabet/Phonetic/NATO.pm: updated the "See Also" section

	* lib/Lingua/Alphabet/Phonetic.pm: updated the "See Also" section

2006-10-06    <Daddy@C-3PO>

	* Makefile.PL: typo

2006-10-01    <Daddy@C-3PO>

	* Code/phonetize-code.PL: FIX spacing between words in certain cases

2004-02-10  Kingpin  <mthurn@verizon.net>

	* t/pod.t: new file

2004-02-06  Kingpin  <mthurn@verizon.net>

	* Code/phonetize-code.PL: check for undef arg
	* lib/Lingua/Alphabet/Phonetic.pm (enunciate): check for undef arg on enunciate()

2004-01-26  Kingpin  <mthurn@verizon.net>

	* Code/phonetize-code.PL: new file

2004-01-10  Kingpin  <mthurn@verizon.net>

	* lib/Lingua/Alphabet/Phonetic.pm: FIRST RELEASE



Code/phonetize-code.PL
exit 0;
# The above line will be stripped off during `make`
#
#line 4

=head1 NAME

phonetize - print a string in a phonetic alphabet for easy reading

=head1 SYNOPSIS

  phonetize Martin

=head1 DESCRIPTION

Takes one argument, a "string".
Converts each character of "string" to its equivalent "word" in the Phonetic Alphabet.
Prints these "words", three "words" per line, one space between "words"

In other words, read the output aloud to your military friends and
they'll know exactly what you're spelling.

=head1 TO DO

=over

=item Let name of phonetic alphabet be an argument (in case more alphabets get implemented)?

=item Handle puncuation somehow?

=item Deal with capital vs. lowercase letters somehow?

=item Let the "three" in "three words per line" be an argument?

=back

=head1 AUTHOR

Martin Thurn

=cut

# use Getopt::Long; # later
use Lingua::Alphabet::Phonetic;

use strict;
use warnings;

use vars qw( $VERSION );
$VERSION = do { my @r = (q$Revision: 1.4 $ =~ /\d+/g); sprintf "%d."."%03d" x $#r, @r };

my $s = shift() || '';
# Insert spaces between every letter:
$s = join(' ', split('', $s));
# Insert {CR} after every third letter:
$s =~ s!(. . .) !$1\n!g;
my $oMilSpeaker = new Lingua::Alphabet::Phonetic('NATO');
my @asMilSpeak = $oMilSpeaker->enunciate($s);
$, = '';
print @asMilSpeak, "\n";
exit 0;

__END__



lib/Lingua/Alphabet/Phonetic/NATO.pm

# $Id: NATO.pm,v 1.6 2008/06/18 01:04:29 Martin Exp $

=head1 NAME

Lingua::Alphabet::Phonetic::NATO - map ABC's to the NATO phonetic letter names

=head1 SYNOPSIS

This is a specialization of Lingua::Alphabet::Phonetic.
You should not use this module;
all interaction should be done with an object of type Lingua::Alphabet::Phonetic.

=head1 NOTES

=head1 SEE ALSO

http://en.wikipedia.org/wiki/NATO_phonetic_alphabet

=head1 BUGS

Please tell the author if you find any!

=head1 LICENSE

This software is released under the same license as Perl itself.

=head1 AUTHOR

Martin Thurn (mthurn@cpan.org).

=cut

#####################################################################

package Lingua::Alphabet::Phonetic::NATO;

use strict;

use base 'Lingua::Alphabet::Phonetic';
our
$VERSION = sprintf("%d.%02d", q$Revision: 1.6 $ =~ /(\d+)\.(\d+)/o);

my @asAlphabet = qw(
Alpha Bravo Charlie Delta Echo Foxtrot Golf Hotel India Juliet Kilo Lima
Mike November Oscar Papa Quebec Romeo Sierra Tango Uniform Victor Whiskey
Xray Yankee Zulu
Zero One Two Three Four Five Six Seven Eight Niner
);
my %hash = map { $_ => shift @asAlphabet } ('a'..'z', 0..9);

sub _name_of_letter
  {
  my $self = shift;
  my $s = shift;
  # print STDERR " + L::A::P::NATO::_name_of_letter($s)\n";
  # If we get more than one character, ignore the rest:
  my $c = lc substr($s, 0, 1);
  if (exists($hash{$c}))
    {
    return $hash{$c};
    } # if
  return $self->SUPER::_name_of_letter($s);
  } # _name_of_letter

1;

__END__



lib/Lingua/Alphabet/Phonetic.pm

# $Id: Phonetic.pm,v 1.8 2008/06/18 01:04:41 Martin Exp $

=head1 NAME

Lingua::Alphabet::Phonetic - map ABC's to phonetic alphabets

=head1 SYNOPSIS

  use Lingua::Alphabet::Phonetic;
  my $oMilSpeaker = new Lingua::Alphabet::Phonetic('NATO');
  my @asMilSpeak = $oMilSpeaker->enunciate('ABC');

=head1 DESCRIPTION

At present, the only alphabet available for conversion is the
U.S. Military / NATO standard where "ABC...YZ" is pronounced "Alpha
Bravo Charlie ... Yankee Zulu".  It is called 'NATO' and it is
included with this distribution.

=head1 METHODS

=over

=cut

#####################################################################

package Lingua::Alphabet::Phonetic;

use strict;

our
$VERSION = sprintf("%d.%02d", q$Revision: 1.8 $ =~ /(\d+)\.(\d+)/o);

=item new

Create an object of this class.  See SYNOPSIS above.

=cut

sub new
  {
  my $class = shift;
  my $sAlphabet = shift || '';
  my $sSubclass = "${class}::$sAlphabet";
  eval "use $sSubclass";
  Carp::croak("Unknown phonetic alphabet $sAlphabet: $@") if ($@);
  my $self = bless {
                   }, $sSubclass;
  return $self;
  } # new


=item enunciate

Given a string, returns a list of phonetic alphabet "words", one word
per character of the original string.  If there is no "word" in the
alphabet for a character, that character is returned in the list
position.

=cut

sub enunciate
  {
  my $self = shift;
  my $s = shift || '';
  my @ac = split('', $s);
  return map { $self->_name_of_letter($_) } @ac;
  } # enunciate


sub _name_of_letter
  {
  # This is the default fallback character --> word mapping.
  my $self = shift;
  my $s = shift;
  # Just return our argument unchanged:
  return $s;
  } # _name_of_letter

=back

=head1 OTHER ALPHABETS

To create a conversion scheme for another alphabet, simply subclass
this module and provide a method _name_of_letter() which takes a
character and returns its phonetic name.  See NATO.pm as an example.

=head1 SEE ALSO

http://en.wikipedia.org/wiki/Spelling_alphabet is a brief overview

http://www.bckelk.uklinux.net/phon.full.html contains a list of
phonetic alphabets from all over the world!

=head1 TO-DO

=over

=item Implement more alphabets.

=item Investigate how we might handle non-ASCII alphabets.  Unicode?

=back

=head1 BUGS

Please tell the author if you find any!

=head1 LICENSE

This software is released under the same license as Perl itself.

=head1 AUTHOR

Martin Thurn (mthurn@cpan.org).

=cut

1;

__END__

